message-broker/src/lib.rs

```diff
@@ -127,53 +126,154 @@ decl_module! {
+			// those are processed after the block initialization phase. Therefore, we have to be
+			// content only with the configuration as per the previous block. That means that
+			// the configuration can be either stale (or be abscent altogether in case of the
+			// beginning of the chain).
+			//
+			// In order to mitigate this, we do the following. At the time, we are only concerned
+			// about `hrmp_max_message_num_per_candidate`. We reserve the amount of weight to process
+			// the number of HRMP messages according to the potentially stale configuration. In
+			// `on_finalize` we will process only the maximum between the announced number of messages
+			// and the actual received in the fresh configuration.
+			//
+			// In the common case, they will be the same. In the case the actual value is smaller
+			// than the announced, we would waste some of weight. In the case the actual value is
+			// greater than the announced, we will miss opportunity to send a couple of messages.
+			weight += T::DbWeight::get().reads_writes(1, 1);
+			let hrmp_max_message_num_per_candidate =
+				<cumulus_parachain_upgrade::Module<T>>::host_configuration()
+					.map(|cfg| cfg.hrmp_max_message_num_per_candidate)
+					.unwrap_or(0);
+			AnnouncedHrmpMessagesPerCandidate::put(hrmp_max_message_num_per_candidate);
+
+			// NOTE that the actual weight consumed by `on_finalize` may turn out lower.
 			weight += T::DbWeight::get().reads_writes(
-				2 + HRMP_MSG_NUM_PER_CANDIDATE as u64,
-				4 + HRMP_MSG_NUM_PER_CANDIDATE as u64,
+				3 + hrmp_max_message_num_per_candidate as u64,
+				4 + hrmp_max_message_num_per_candidate as u64,
 			);
 
 			weight
 		}
 
 		fn on_finalize() {
+			let host_config = <cumulus_parachain_upgrade::Module<T>>::host_configuration().unwrap();
+			let relevant_messaging_state = <cumulus_parachain_upgrade::Module<T>>::relevant_messaging_state().unwrap();
+
 			<Self as Store>::PendingUpwardMessages::mutate(|up| {
-				let num = cmp::min(UMP_MSG_NUM_PER_CANDIDATE, up.len());
-				storage::unhashed::put(
-					well_known_keys::UPWARD_MESSAGES,
-					&up[0..num],
+				let (count, size) = relevant_messaging_state.relay_dispatch_queue_size;
+
+				let available_capacity = cmp::min(
+					host_config.max_upward_queue_count.saturating_sub(count),
+					host_config.max_upward_message_num_per_candidate,
 				);
+				let available_size = host_config.max_upward_queue_size.saturating_sub(size);
+
+				// Count the number of messages we can possibly fit in the given constraints, i.e.
+				// available_capacity and available_size.
+				let num = up
+					.iter()
+					.scan(
+						(available_capacity as usize, available_size as usize),
+						|(cnt, size), msg| match (cnt.checked_sub(1), size.checked_sub(msg.len())) {
+							(Some(cnt), Some(size)) => Some((cnt, size)),
+							_ => None,
+						},
+					)
+					.count();
+
+				// TODO: #274 Return back messages that do not longer fit into the queue.
```

Comment:
By that you mean messages that are now bigger than the maximum allowed message size?

Reply:
Why do we even support to change the max message size for an open channel?

Reply:
Ahh that is an upward message, this makes somewhat sense.